bindings/cpu_profiler.cc

```diff
@@ -52,6 +57,171 @@ v8::CpuProfilingLoggingMode GetLoggingMode() {
   return kDefaultLoggingMode;
 }
 
+class SentryProfile;
+
+enum class ProfileStatus {
+  kNotStarted,
+  kStarted,
+  kStopped,
+};
+
+class HeapStatisticsTimer {
+private:
+  uint64_t period_ms;
+  std::unordered_map<std::string, std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)>> listeners;
+  v8::Isolate* isolate;
+  v8::HeapStatistics heap_stats;
+
+public:
+  uv_timer_t timer;
+
+  HeapStatisticsTimer(uv_loop_t* loop) :
+    period_ms(100),
+    listeners(std::unordered_map<std::string, std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)>>()),
+    isolate(v8::Isolate::GetCurrent()),
+    heap_stats(v8::HeapStatistics())
+  {
+    uv_timer_init(loop, &timer);
+    timer.data = this;
+  };
+
+  static void callback(uv_timer_t* handle);
+  static void add_listener(uv_timer_t* handle, const char* profile_id, std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)> cb);
+  static void remove_listener(uv_timer_t* handle, const char* profile_id, std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)>& cb);
+};
+
+void HeapStatisticsTimer::callback(uv_timer_t* handle) {
+  HeapStatisticsTimer* self = static_cast<HeapStatisticsTimer*>(handle->data);
+  self->isolate->GetHeapStatistics(&self->heap_stats);
+  uint64_t ts = uv_hrtime();
+
+  for (auto& listener : self->listeners) {
+    listener.second(handle, ts, self->heap_stats);
+  }
+}
+
+void HeapStatisticsTimer::add_listener(uv_timer_t* handle, const char* profile_id, std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)> cb) {
+  auto self = static_cast<HeapStatisticsTimer*>(handle->data);
+  self->listeners.emplace(std::string(profile_id), cb);
+
+  if (self->listeners.size() == 1) {
+    uv_timer_set_repeat(&self->timer, self->period_ms);
+    uv_timer_start(&self->timer, callback, 0, self->period_ms);
+  }
+}
+
+void HeapStatisticsTimer::remove_listener(uv_timer_t* handle, const char* profile_id, std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)>& cb) {
+  auto self = static_cast<HeapStatisticsTimer*>(handle->data);
+  self->listeners.erase(std::string(profile_id));
+
+  if (self->listeners.size() == 0) {
+    uv_timer_stop(&self->timer);
+  }
+};
+
+class Profiler {
+public:
+  std::unordered_map<std::string, SentryProfile*> active_profiles = std::unordered_map<std::string, SentryProfile*>();
+
+  v8::CpuProfiler* cpu_profiler;
+  HeapStatisticsTimer heap_statistics_timer = HeapStatisticsTimer(uv_default_loop());
```

Comment:
this could be initialized as part of your explicit initializer on line 130 to keep things in one place, you would just add `heap_statistics_timer(uv_default_loop())` and remove the assignment here